File: icpc/greater-ny-2023/problems/F-forestforthetrees/forest-zack.cpp
// Zack tries to remember C++.
#include <unordered_set>
#include <iostream>
#include <utility>
#define MAXW 200000000
#define HALFW 100000000
using namespace std;

typedef unordered_set<long> treeset;

long loc(long px, long py) {
  return MAXW*(py+HALFW)+(px+HALFW);
}

long txl(long t) {
  return t%MAXW - HALFW;
}

long tyl(long t) {
  return t/MAXW - HALFW;
}

bool checkpos(long px, long py, treeset found, treeset trees, int maxr) {
  long l = loc(px,py);
  for (auto& t : trees) {
    if ((found.count(t) == 0) && (abs(px-txl(t))+abs(py-tyl(t)) <= maxr)) {
      //cout << "missing" << t << endl;
      return false;
    }
  }
  return true;
}

int main(int argc, char** argv) {
  int nt, ns, maxr;
  cin >> nt >> ns >> maxr;
  treeset trees;
  int tx, ty;
  for (int tn = 0; tn < nt; tn++) {
    cin >> tx >> ty;
    trees.insert(loc(tx,ty));
  }
  //cout << trees.size() << "trees";
  // sensor-x sensor-y (sideways amt, forward amt)
  int sens[ns][2];
  // int sx,sy;
  for (int s = 0; s < ns; s++) {
    cin >> sens[s][0] >> sens[s][1];
    if (sens[s][0] == 0 && sens[s][1] == 0) {
      cout << "Impossible" << endl;
      exit(0);
    }
  }
  bool hasplace = false;
  pair<int,int> place;
  
  // pointing +x
  for (auto& t : trees) {
    long px = txl(t)-sens[0][1];
    long py = tyl(t)+sens[0][0];
    // cout << "pos " << px << " " << py << endl;
    bool fail = false;
    treeset found;
    found.insert(t);
    for (int s = 1; s < ns; s++) {
      long tx = px+sens[s][1];
      long ty = py-sens[s][0];
      // cout << tx << " " << ty << endl;
      long tloc = loc(tx,ty);
      if (trees.count(tloc) > 0) {
        found.insert(tloc);
      } else {
        // cout << "not " << tx << " " << ty << endl;
        fail = true;
        break;
      }
    }
    if (!fail && checkpos(px,py,found,trees,maxr)) {
      //cout << "place " << px << " " << py << " +x" << endl;
      if (!hasplace) {
        place = make_pair(px,py);
        hasplace = true;
      }
      else {
        cout << "Ambiguous" << endl;
        exit(0);
      }
    }
  }

  // pointing +y
  for (auto& t : trees) {
    long px = txl(t)-sens[0][0];
    long py = tyl(t)-sens[0][1];
    bool fail = false;
    treeset found;
    found.insert(t);
    for (int s = 1; s < ns; s++) {
      long tx = px+sens[s][0];
      long ty = py+sens[s][1];
      long tloc = loc(tx,ty);
      if (trees.count(tloc) > 0) {
        found.insert(tloc);
      } else {
        fail = true;
        break;
      }
    }
    if (!fail && checkpos(px,py,found,trees,maxr)) {
      //cout << "place " << px << " " << py << " +y" << endl;
      if (!hasplace) {
        place = make_pair(px,py);
        hasplace = true;
      }
      else {
        cout << "Ambiguous" << endl;
        exit(0);
      }
    }
  }

  // pointing -x
  for (auto& t : trees) {
    long px = txl(t)+sens[0][1];
    long py = tyl(t)-sens[0][0];
    //cout << "pos " << px << " " << py << endl;
    bool fail = false;
    treeset found;
    found.insert(t);
    for (int s = 1; s < ns; s++) {
      long tx = px-sens[s][1];
      long ty = py+sens[s][0];
      long tloc = loc(tx,ty);
      if (trees.count(tloc) > 0) {
        found.insert(tloc);
      } else {
        fail = true;
        break;
      }
    }
    if (!fail && checkpos(px,py,found,trees,maxr)) {
      //cout << "place " << px << " " << py << " -x" << endl;
      if (!hasplace) {
        place = make_pair(px,py);
        hasplace = true;
      }
      else {
        cout << "Ambiguous" << endl;
        exit(0);
      }
    }
  }

  // pointing -y
  for (auto& t : trees) {
    long px = txl(t)+sens[0][0];
    long py = tyl(t)+sens[0][1];
    bool fail = false;
    treeset found;
    found.insert(t);
    for (int s = 1; s < ns; s++) {
      long tx = px-sens[s][0];
      long ty = py-sens[s][1];
      long tloc = loc(tx,ty);
      if (trees.count(tloc) > 0) {
        found.insert(tloc);
      } else {
        fail = true;
        break;
      }
    }
    if (!fail && checkpos(px,py,found,trees,maxr)) {
      //cout << "place " << px << " " << py << " -y" << endl;
      if (!hasplace) {
        place = make_pair(px,py);
        hasplace = true;
      }
      else {
        cout << "Ambiguous" << endl;
        exit(0);
      }
    }
  }

  if (!hasplace)
    cout << "Impossible" << endl;
  else
    cout << place.first << " " << place.second << endl;
}

  

File: icpc/greater-ny-2023/problems/F-forestforthetrees/forest-zack.py
# discrete forest for the trees
# pretty naive
# also very sorry for the code repetition 
# zbutler

import sys

def checkpos(px,py,found,trees,maxr):
    for t in trees:
        if t not in found and (abs(px-t[0])+abs(py-t[1])) <= maxr:
            #print('missing',t)
            return False
    return True

#MAXR = 10000
# #trees #sensed-trees
nt, ns, maxr = [int(x) for x in input().strip().split()]
if ns == 0:
    print("Ambiguous")
    sys.exit(0)
trees = set()
# tree-x tree-y
for _ in range(nt):
    trees.add(tuple([int(x) for x in input().strip().split()]))
# sensor-x sensor-y (sideways amt, forward amt)
sens = []
for _ in range(ns):
    sens.append(tuple([int(x) for x in input().strip().split()]))
    #if sens[-1] == (0,0):
    #    print("Impossible")
    #    sys.exit(0)
        
place = None
# pointing +x
for t in trees:
    px = t[0]-sens[0][1]
    py = t[1]+sens[0][0]
    #print(px,py)
    if (px,py) in trees:
        continue
    fail = False
    found = set()
    found.add(t)
    for s in range(1,ns):
        tx = px+sens[s][1]
        ty = py-sens[s][0]
        #print(tx,ty)
        if (tx,ty) in trees:
            found.add((tx,ty))
        else:
            #print('not',tx,ty)
            fail = True
            break
    if not fail and checkpos(px,py,found,trees,maxr):
        if place is None:
            place = (px,py)
        else:
            print("Ambiguous")
            sys.exit(0)

# pointing +y
for t in trees:
    px = t[0]-sens[0][0]
    py = t[1]-sens[0][1]
    if (px,py) in trees:
        continue
    #print(px,py)
    fail = False
    found = set()
    found.add(t)
    for s in range(1,ns):
        tx = px+sens[s][0]
        ty = py+sens[s][1]
        #print(tx,ty)
        if (tx,ty) in trees:
            found.add((tx,ty))
        else:
            #print('not',tx,ty)
            fail = True
            break
    if not fail and checkpos(px,py,found,trees,maxr):
        if place is None:
            place = (px,py)
        else:
            print("Ambiguous")
            sys.exit(0)

# pointing -x
for t in trees:
    px = t[0]+sens[0][1]
    py = t[1]-sens[0][0]
    if (px,py) in trees:
        continue
    #print(px,py)
    fail = False
    found = set()
    found.add(t)
    for s in range(1,ns):
        tx = px-sens[s][1]
        ty = py+sens[s][0]
        #print(tx,ty)
        if (tx,ty) in trees:
            found.add((tx,ty))
        else:
            #print('not',tx,ty)
            fail = True
            break
    if not fail and checkpos(px,py,found,trees,maxr):
        if place is None:
            place = (px,py)
        else:
            print("Ambiguous")
            sys.exit(0)

# pointing -y
for t in trees:
    px = t[0]+sens[0][0]
    py = t[1]+sens[0][1]
    if (px,py) in trees:
        continue
    #print(px,py)
    fail = False
    found = set()
    found.add(t)
    for s in range(1,ns):
        tx = px-sens[s][0]
        ty = py-sens[s][1]
        #print(tx,ty)
        if (tx,ty) in trees:
            found.add((tx,ty))
        else:
            #print('not',tx,ty)
            fail = True
            break
    if not fail and checkpos(px,py,found,trees,maxr):
        if place is None:
            place = (px,py)
        else:
            print("Ambiguous")
            sys.exit(0)

if not place:
    print("Impossible")
else:
    print(place[0],place[1])
